include/revng/Pipeline/Containers.h

```diff
@@ -0,0 +1,227 @@
+  Containers() = default;
+  Containers(Containers &&) = default;
+  Containers &operator=(Containers &&) = default;
+  ~Containers() = default;
+  Containers(const Containers &) = delete;
+  Containers &operator=(const Containers &) = delete;
+
+  Containers cloneFiltered(const ContainersStatus &Targets);
+
+  void mergeContainers(Containers &&Other) {
+    for (auto &Entry : Other.Content) {
+      revng_assert(Content.count(Entry.first()) != 0);
+      MaybeContainer &LContainer = Content.find(Entry.first())->second;
+      MaybeContainer &RContainer = Entry.second;
+      LContainer.mergeBack(std::move(RContainer));
+    }
+  }
+
+  template<typename ContainerDerivedType>
+  ContainerDerivedType &getOrCreate(llvm::StringRef Name) {
+    revng_assert(containsOrCanCreate(Name));
+    auto &Base = Content.find(Name)->second.getOrCreate();
+    revng_assert(llvm::isa<ContainerDerivedType>(Base));
+    return llvm::cast<ContainerDerivedType>(Base);
+  }
+
+  template<typename ContainerDerivedType>
+  const ContainerDerivedType &get(llvm::StringRef Name) const {
+    revng_assert(contains(Name));
+    const auto &Base = Content.find(Name)->second.get();
+    revng_assert(llvm::isa<ContainerDerivedType>(Base));
+    return llvm::cast<ContainerDerivedType>(Base);
+  }
+
+  ContainerBase &get(llvm::StringRef Name) {
+    revng_assert(contains(Name));
+    return Content.find(Name)->second.getOrCreate();
+  }
+
+  const ContainerBase &get(llvm::StringRef Name) const {
+    revng_assert(contains(Name));
+    return Content.find(Name)->second.get();
+  }
+
+  bool contains(llvm::StringRef Name) const {
+    return Content.count(Name) != 0 and Content.find(Name)->second.hasValue();
+  }
+
+  bool containsOrCanCreate(llvm::StringRef Name) const {
+    return Content.count(Name) != 0;
+  }
+
+  bool contains(const Target &Target) const;
+
+  void addEmpty(llvm::StringRef Name, ContainerFactory &Factory) {
+    Content.insert_or_assign(Name, MaybeContainer(Factory));
+  }
+
+  void add(llvm::StringRef Name, MaybeContainer Container) {
+    Content.insert_or_assign(Name, std::move(Container));
+  }
+
+  void addNonEmpty(llvm::StringRef Name,
+                   ContainerFactory &Factory,
+                   std::unique_ptr<ContainerBase> Container) {
+    Content.insert_or_assign(Name,
+                             MaybeContainer(Factory, std::move(Container)));
+  }
+
+  bool isContainerEmpty(llvm::StringRef ContainerName) const {
+    return Content.find(ContainerName)->second.hasValue();
+  }
+
+  llvm::Error remove(const ContainersStatus &ToRemove);
+  void intersect(ContainersStatus &ToIntersect) const;
+
+  llvm::Error store(llvm::StringRef DirectoryPath) const;
+  llvm::Error load(llvm::StringRef DirectoryPath);
+
+  llvm::Expected<const ContainerBase *>
+  getContainer(llvm::StringRef ContainerName) const;
+
+  llvm::Expected<ContainerBase *>
+  getOrCreateContainer(llvm::StringRef ContainerName);
```

Comment:
Too many `get` with unclear semantics. Can't we just have `.find`, `operator[]` and `at`?

Reply:
ok i fixed this at this level of abstraction, but at higher levels (like runner) we can't really use stuf like operator[] because it has 2 arguments.